integration/benchmark_test.go

```diff
@@ -0,0 +1,215 @@
+		expectedOS:    analyzer.OS{Name: "1.0", Family: "photon"},
+	},
+	{
+		name:          "happy path, registry.redhat.io/ubi7",
+		imageName:     "registry.redhat.io/ubi7",
+		imageFile:     "testdata/fixtures/ubi-7.tar.gz",
+		expectedFiles: []string{"etc/redhat-release", "etc/system-release", "/config", "var/lib/rpm/Packages", "etc/os-release"},
+		expectedOS:    analyzer.OS{Name: "7.7", Family: "redhat"},
+	},
+	{
+		name:          "happy path, opensuse leap 15.1",
+		imageName:     "opensuse/leap:latest",
+		imageFile:     "testdata/fixtures/opensuse-leap-151.tar.gz",
+		expectedFiles: []string{"usr/lib/os-release", "usr/lib/sysimage/rpm/Packages", "/config", "etc/os-release"},
+		expectedOS:    analyzer.OS{Name: "15.1", Family: "opensuse.leap"},
+	},
+	{
+		name:                 "happy path, vulnimage with lock files",
+		imageName:            "knqyf263/vuln-image:1.2.3",
+		imageFile:            "testdata/fixtures/vulnimage.tar.gz",
+		expectedFiles:        []string{"etc/os-release", "node-app/package-lock.json", "python-app/Pipfile.lock", "ruby-app/Gemfile.lock", "rust-app/Cargo.lock", "/config", "etc/alpine-release", "lib/apk/db/installed", "php-app/composer.lock"},
+		expectedOS:           analyzer.OS{Name: "3.7.1", Family: "alpine"},
+		expectedLibraries:    "testdata/goldens/knqyf263vuln-image:1.2.3.expectedlibs.golden",
+		expectedPkgsFromCmds: "testdata/goldens/knqyf263vuln-image:1.2.3.expectedpkgsfromcmds.golden",
+	},
+}
+
+// benchCache is shared across benchmarks
+var benchCache string
+
+func runChecksBench(b *testing.B, ac analyzer.Config, ctx context.Context, tc testCase) {
+	for i := 0; i < b.N; i++ {
+		actualFiles, err := ac.Analyze(ctx, tc.imageFile)
+		require.NoError(b, err)
+
+		osFound, err := analyzer.GetOS(actualFiles)
+		require.NoError(b, err)
+
+		_, err = analyzer.GetPackages(actualFiles)
+		require.NoError(b, err)
+
+		_, err = analyzer.GetPackagesFromCommands(osFound, actualFiles)
+		require.NoError(b, err)
+
+		_, err = analyzer.GetLibraries(actualFiles)
+	}
+}
+
+func BenchmarkFanal_Library_DockerMode_WithoutCache(b *testing.B) {
+	benchCache, _ = ioutil.TempDir("", "TestFanal_Library_*")
+	for _, tc := range testCases {
+		tc := tc // save a copy of tc for use in t.Run https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721
+		var ctx context.Context
+		var ac analyzer.Config
+
+		ctx, _, _, ac = setup(b, tc, benchCache)
+		b.Run(tc.name, func(b *testing.B) {
+			b.ReportAllocs()
+			b.ResetTimer()
+			runChecksBench(b, ac, ctx, tc)
+			b.StopTimer()
+		})
+	}
+}
+
+func BenchmarkFanal_Library_DockerMode_WithCache(b *testing.B) {
+	defer os.RemoveAll(benchCache)
+
+	for _, tc := range testCases {
+		tc := tc // save a copy of tc for use in t.Run https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721
+		// setup
+		opt := types.DockerOption{
+			Timeout:  600 * time.Second,
+			SkipPing: true,
+		}
+
+		// resuse cache created in previous test
+		c := cache.Initialize(benchCache)
```

Comment:
We should avoid reusing the state in a previous test.

> Independent: Tests should not depend on the state of the previous test
https://dzone.com/articles/writing-your-first-unit-tests

This function depends on the order of benchmarks.

Reply:
How about the following way?

1. We define a function to run `Analyze`, `GetOS`, etc.
```
func run(b *testing.B, ctx context.Context,ac analyzer.Config, imageFile string) {
		actualFiles, err := ac.Analyze(ctx, imageFile)
		require.NoError(b, err)

		osFound, err := analyzer.GetOS(actualFiles)
		require.NoError(b, err)

		_, err = analyzer.GetPackages(actualFiles)
		require.NoError(b, err)

		_, err = analyzer.GetPackagesFromCommands(osFound, actualFiles)
		require.NoError(b, err)

		_, err = analyzer.GetLibraries(actualFiles)
		require.NoError(b, err)
```

Next, `runChecksBench` uses it.

```
func runChecksBench(b *testing.B, ac analyzer.Config, ctx context.Context, tc testCase) {
    for i := 0; i < b.N; i++ {
        run(b, ctx, ac, tc.imageFile)
    }
}

```
		b.Run(tc.name, func(b *testing.B) {
			b.ReportAllocs()

                         // generate cache before measuring a benchmark
                         run(b, ctx, ac, tc.imageFile)

			b.ResetTimer()
			runChecksBench(b, ac, context.Background(), tc)
			b.StopTimer()
		})
```

Reply:
Excellent idea. I implemented this: https://github.com/aquasecurity/fanal/pull/72/commits/e26a4093de03bc069e0a6430681d38578b57c49c